packages/turbo/ios/RNSession.swift

```diff
@@ -11,61 +11,90 @@ import WebKit
-    
-  override init() { }
-    
-  init(eventEmitter: RCTEventEmitter, sessionHandle: NSString){
-    self.eventEmitter = eventEmitter
+  private var visitableViews: [RNSessionSubscriber] = []
+  private var sessionHandle: NSString
+  private var webViewConfiguration: WKWebViewConfiguration
+  
+  init(sessionHandle: NSString, webViewConfiguration: WKWebViewConfiguration) {
     self.sessionHandle = sessionHandle
+    self.webViewConfiguration = webViewConfiguration
   }
-    
-  convenience init(eventEmitter: RCTEventEmitter, sessionHandle: NSString, applicationNameForUserAgent: NSString?){
-    self.init(eventEmitter: eventEmitter, sessionHandle: sessionHandle)
-    self.applicationNameForUserAgent = applicationNameForUserAgent as String?
-  }
-  
+
   public lazy var turboSession: Session = {
-    let configuration = WKWebViewConfiguration()
-    configuration.userContentController.add(self, name: "nativeApp")
-    configuration.userContentController.add(RNStradaWKUserContentController(eventEmitter: eventEmitter), name: "stradaNative")
-    configuration.applicationNameForUserAgent = applicationNameForUserAgent
-    return Session(webViewConfiguration: configuration)
+    webViewConfiguration.userContentController.add(self, name: "nativeApp")
+    let session = Session(webViewConfiguration: webViewConfiguration)
+    session.delegate = self
+    return session
   }()
+  public lazy var webView: WKWebView = turboSession.webView
   
-  func registerVisitableView(newView: SessionSubscriber) {
-    if (!registeredVisitableViews.contains {
+  func registerVisitableView(newView: RNSessionSubscriber) {
+    if (!visitableViews.contains {
       $0.id == newView.id
     }) {
-      registeredVisitableViews.append(newView)
+      visitableViews.append(newView)
     }
-    newView.attachDelegateAndVisit(newView.controller!)
   }
   
-  func removeVisitableView(view: SessionSubscriber) {
-    // New view is not registered when presentation is modal
-    let isViewModal = registeredVisitableViews.last?.id == view.id
-    let viewIdx = registeredVisitableViews.lastIndex(where: {
+  func unregisterVisitableView(view: RNSessionSubscriber) {
+    let wasTopMostView = visitableViews.last?.id == view.id
+    
+    let viewIdx = visitableViews.lastIndex(where: {
       view.id == $0.id
     })
-    registeredVisitableViews.remove(at: viewIdx!)
-    guard let newView = registeredVisitableViews.last else {
-      return
-    }
-    if (isViewModal) {
-      newView.attachDelegateAndVisit(newView.controller!)
+    visitableViews.remove(at: viewIdx!)
+
+    // The new top-most view is not registered when the previous top-most view is a modal
+    if (wasTopMostView) {
+      guard let newView = visitableViews.last else {
+        return
+      }
+      visitableViewWillAppear(view: newView)
     }
   }
+    
+  func visitableViewWillAppear(view: RNSessionSubscriber) {
+    turboSession.visitableViewWillAppear(view.controller)
+  }
```

Comment:
I'm wondering if it makes sense to have a method here, that is only used once. I don't think `visitableViewWillAppear` would be called anywhere else than from `unregisterVisitableView`?

One other thing to consider is if we need this at all - if modals are supposed to have different session handles, the scenario where we'd need to re-register that the view behind a closing modal becomes the active one, isn't relevant.

Reply:
>I'm wondering if it makes sense to have a method here, that is only used once. I don't think visitableViewWillAppear would be called anywhere else than from unregisterVisitableView?

I've decided to wrap this into the separate method, because I think it is more readable when you call `visitableViewWillAppear(view: newView)` rather than calling a method directly on `turboSession` and passing a `controller` property of the new view.

> One other thing to consider is if we need this at all - if modals are supposed to have different session handles, the scenario where we'd need to re-register that the view behind a closing modal becomes the active one, isn't relevant.

Hmm, no strong opinion here, we can either remove it or leave it.